process/block/preprocess/transactionsV2.go

```diff
@@ -52,213 +20,81 @@ func (txs *transactions) createAndProcessMiniBlocksFromMeV2(
+		tx, miniBlock, shouldContinue := txs.shouldContinueProcessingTx(
+			isShardStuck,
+			sortedTxs[index],
+			cpmbi)
+		if !shouldContinue {
 			continue
 		}
 
-		if len(senderAddressToSkip) > 0 {
-			if bytes.Equal(senderAddressToSkip, tx.GetSndAddr()) {
-				processingInfo.numTxsSkipped++
-				continue
-			}
-		}
-
-		_, txTypeDstShard := txs.txTypeHandler.ComputeTransactionType(tx)
-		isReceiverSmartContractAddress := txTypeDstShard == process.SCDeployment || txTypeDstShard == process.SCInvoking
-		firstMiniBlockSplitForReceiverShardFound := (isReceiverSmartContractAddress && mapScsForShard[receiverShardID] == 0 && mapTxsForShard[receiverShardID] > 0) ||
-			(!isReceiverSmartContractAddress && mapTxsForShard[receiverShardID] == 0 && mapScsForShard[receiverShardID] > 0)
-
-		miniBlock, ok := mapMiniBlocks[receiverShardID]
-		if !ok {
-			log.Debug("mini block is not created", "shard", receiverShardID)
-			continue
-		}
-
-		numNewMiniBlocks := 0
-		if len(miniBlock.TxHashes) == 0 || firstCrossShardScCallOrSpecialTxFound {
-			numNewMiniBlocks = 1
-		}
-		numNewTxs := 1
+		txHash := sortedTxs[index].TxHash
+		senderShardID := sortedTxs[index].SenderShardID
+		receiverShardID := sortedTxs[index].ReceiverShardID
 
-		isCrossShardScCallOrSpecialTx := receiverShardID != txs.shardCoordinator.SelfId() &&
-			(isReceiverSmartContractAddress || len(tx.RcvUserName) > 0)
-		if isCrossShardScCallOrSpecialTx {
-			if !firstCrossShardScCallOrSpecialTxFound {
-				numNewMiniBlocks++
-			}
-			if mapCrossShardScCallsOrSpecialTxs[receiverShardID] >= maxCrossShardScCallsOrSpecialTxsPerShard {
-				numNewTxs += core.AdditionalScrForEachScCallOrSpecialTx
-			}
-		}
+		isMiniBlockEmpty := len(miniBlock.TxHashes) == 0
+		txMbInfo := txs.getTxAndMbInfo(
+			tx,
+			isMiniBlockEmpty,
+			receiverShardID,
+			cpmbi)
 
-		if isMaxBlockSizeReached(numNewMiniBlocks, numNewTxs) {
+		if isMaxBlockSizeReached(txMbInfo.numNewMiniBlocks, txMbInfo.numNewTxs) {
 			log.Debug("max txs accepted in one block is reached",
-				"num txs added", processingInfo.numTxsAdded,
+				"num txs added", cpmbi.processingInfo.numTxsAdded,
 				"total txs", len(sortedTxs))
 			break
 		}
 
-		addressHasEnoughBalance, isAddressSet, txMaxTotalCost := txs.hasAddressEnoughInitialBalance(tx)
-		if !addressHasEnoughBalance {
-			processingInfo.numTxsWithInitialBalanceConsumed++
-			continue
-		}
-
-		txType := nonScTx
-		if isReceiverSmartContractAddress {
-			txType = scTx
-		}
-
 		err := txs.processTransaction(
 			tx,
-			txType,
+			txMbInfo.txType,
 			txHash,
 			senderShardID,
 			receiverShardID,
-			&gasInfo,
-			mapGasConsumedByMiniBlockInReceiverShard,
-			&processingInfo)
-		if err != nil && !errors.Is(err, process.ErrFailedTransaction) {
-			if errors.Is(err, process.ErrHigherNonceInTransaction) {
-				senderAddressToSkip = tx.GetSndAddr()
-			}
-			continue
-		}
-
-		senderAddressToSkip = []byte("")
-
-		gasRefunded := txs.gasHandler.GasRefunded(txHash)
```

Comment:
This gasRefunded approach should be modified exactly like in PR: https://github.com/ElrondNetwork/elrond-go/pull/3016